File: src/application/usecases/searches/anime-command-usecase.ts
import { Anime } from "@/infrastructure/jikan/dtos/anime-search-response-dto";
import { TranslationService } from "@/infrastructure/i18n/translation-service";
import { JikanService } from "@/infrastructure/jikan/jikan-service";
import { ChatInputCommandInteraction, EmbedBuilder } from "discord.js";
import { inject, injectable } from "inversify";
import { AbstractBaseUseCase } from "../base-usecase";

@injectable()
export class AnimeCommandUseCase extends AbstractBaseUseCase<ChatInputCommandInteraction, void> {
  constructor(
    @inject(JikanService) private readonly jikanService: JikanService,
    @inject(TranslationService) private readonly translate: TranslationService
  ) {
    super();
  }

  public async execute(interaction: ChatInputCommandInteraction): Promise<void> {
    const animeName = interaction.options.getString("title", true);

    const animeData = await this.jikanService.getAnimeDetails(animeName, true);
    if (!animeData) {
      await interaction.reply({
        content: this.translate.t("CMD_ANIME_NOT_FOUND", { title: animeName })
      });
      return;
    }

    const embed = this.buildAnimeEmbed(interaction, animeData);
    await interaction.reply({
      embeds: [embed]
    });
  }

  private buildAnimeEmbed(interaction: ChatInputCommandInteraction, animeData: Anime) {
    const defaultValue = this.translate.t("COMMON_NOT_AVAILABLE");
    const synopsis =
      animeData.synopsis.length > 450
        ? animeData.synopsis.substring(0, 450) + "..."
        : animeData.synopsis;
    const genres = animeData.genres
      ? animeData.genres.map((genre) => genre.name).join(", ")
      : defaultValue;
    const studios = animeData.studios
      ? animeData.studios.map((studio) => studio.name).join(", ")
      : defaultValue;

    const embed = new EmbedBuilder()
      .setTitle(animeData.title || defaultValue)
      .setURL(animeData.url || "#")
      .setDescription(
        `**${this.translate.t("CMD_ANIME_DESCRIPTION_ENGLISH_TITLE")}:** ${
          animeData.title_english || defaultValue
        }
        \n**${this.translate.t("CMD_ANIME_DESCRIPTION_SYNOPSIS")}:** ${synopsis}
        \n**${this.translate.t("CMD_ANIME_DESCRIPTION_TRAILER")}:** ${
          animeData.trailer?.url || defaultValue
        }`
      )
      .setColor(0x00ff00)
      .setTimestamp()
      .setFooter({
        text: this.translate.t("COMMON_REQUESTED_BY", { username: interaction.user.username }),
        iconURL: interaction.user.displayAvatarURL()
      })
      .addFields(
        {
          name: this.translate.t("CMD_ANIME_FIELD_EPISODES"),
          value: animeData.episodes?.toString() || defaultValue,
          inline: true
        },
        {
          name: this.translate.t("CMD_ANIME_FIELD_TYPE_STATUS"),
          value: `${animeData.type} | ${animeData.status}`,
          inline: true
        },
        {
          name: this.translate.t("CMD_ANIME_FIELD_SCORE_RANK"),
          value: `${animeData.score || defaultValue} | #${animeData.rank || defaultValue}`,
          inline: true
        },
        {
          name: this.translate.t("CMD_ANIME_FIELD_GENRES"),
          value: genres,
          inline: true
        },
        {
          name: this.translate.t("CMD_ANIME_FIELD_AIRED"),
          value: `${animeData.aired?.prop?.from?.year || defaultValue} | ${animeData.season}`,
          inline: true
        },
        {
          name: this.translate.t("CMD_ANIME_FIELD_STUDIOS"),
          value: studios,
          inline: true
        }
      );

    if (animeData.images?.jpg?.image_url) {
      embed.setImage(animeData.images.jpg.image_url);
    }

    return embed;
  }
}

File: src/main/java/dev/quasemago/aura/application/usecase/searches/AnimeCommandUseCase.java
package dev.quasemago.aura.application.usecase.searches;

import dev.quasemago.aura.infrastructure.i18n.TranslationService;
import dev.quasemago.aura.infrastructure.jikan.JikanService;
import dev.quasemago.aura.infrastructure.jikan.dto.Anime;
import dev.quasemago.aura.shared.DiscordHelpers;
import discord4j.core.event.domain.interaction.ChatInputInteractionEvent;
import discord4j.core.object.entity.User;
import discord4j.core.spec.EmbedCreateSpec;
import discord4j.rest.util.Color;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;

import java.time.Instant;
import java.util.Map;

@Service
public class AnimeCommandUseCase {
    private final JikanService jikanService;
    private final TranslationService translate;

    public AnimeCommandUseCase(JikanService jikanService, TranslationService translate) {
        this.jikanService = jikanService;
        this.translate = translate;
    }

    public Mono<Void> execute(ChatInputInteractionEvent event, User author) {
        var titleOption = translate.t("CMD_ANIME_OPTION_TITLE_NAME");
        return DiscordHelpers.getRequiredOption(event, titleOption)
                .map(option -> option.asString())
                .flatMap(title -> event.deferReply()
                        .then(Mono.fromCallable(() -> jikanService.getAnimeDetails(title, true))
                                .subscribeOn(Schedulers.boundedElastic()))
                        .flatMap(anime -> anime
                                .map(value -> event.editReply()
                                        .withEmbeds(buildAnimeEmbed(author, value))
                                        .then())
                                .orElseGet(() -> event.editReply(translate.t("CMD_ANIME_NOT_FOUND", Map.of("title", title))).then())));
    }

    private EmbedCreateSpec buildAnimeEmbed(User author, Anime anime) {
        var defaultValue = translate.t("COMMON_NOT_AVAILABLE");
        var synopsis = DiscordHelpers.truncate(anime.synopsis(), 450, defaultValue);
        var genres = DiscordHelpers.joinNames(anime.genres(), defaultValue);
        var studios = DiscordHelpers.joinNames(anime.studios(), defaultValue);
        var imageUrl = anime.images() != null && anime.images().jpg() != null
                ? anime.images().jpg().imageUrl()
                : null;
        var trailerUrl = anime.trailer() != null ? anime.trailer().url() : null;
        var airedYear = anime.aired() != null && anime.aired().prop() != null && anime.aired().prop().from() != null
                ? anime.aired().prop().from().year()
                : null;

        var embed = EmbedCreateSpec.builder()
                .title(DiscordHelpers.valueOrDefault(anime.title(), defaultValue))
                .url(DiscordHelpers.valueOrDefault(anime.url(), "#"))
                .description("""
                        **%s:** %s

                        **%s:** %s

                        **%s:** %s""".formatted(
                        translate.t("CMD_ANIME_DESCRIPTION_ENGLISH_TITLE"),
                        DiscordHelpers.valueOrDefault(anime.titleEnglish(), defaultValue),
                        translate.t("CMD_ANIME_DESCRIPTION_SYNOPSIS"),
                        synopsis,
                        translate.t("CMD_ANIME_DESCRIPTION_TRAILER"),
                        DiscordHelpers.valueOrDefault(trailerUrl, defaultValue)
                ))
                .color(Color.of(0x00ff00))
                .timestamp(Instant.now())
                .footer(translate.t("COMMON_REQUESTED_BY", Map.of("username", author.getUsername())), author.getAvatarUrl())
                .addField(translate.t("CMD_ANIME_FIELD_EPISODES"), DiscordHelpers.valueOrDefault(anime.episodes(), defaultValue), true)
                .addField(translate.t("CMD_ANIME_FIELD_TYPE_STATUS"), "%s | %s".formatted(
                        DiscordHelpers.valueOrDefault(anime.type(), defaultValue),
                        DiscordHelpers.valueOrDefault(anime.status(), defaultValue)
                ), true)
                .addField(translate.t("CMD_ANIME_FIELD_SCORE_RANK"), "%s | #%s".formatted(
                        DiscordHelpers.valueOrDefault(anime.score(), defaultValue),
                        DiscordHelpers.valueOrDefault(anime.rank(), defaultValue)
                ), true)
                .addField(translate.t("CMD_ANIME_FIELD_GENRES"), genres, true)
                .addField(translate.t("CMD_ANIME_FIELD_AIRED"), "%s | %s".formatted(
                        DiscordHelpers.valueOrDefault(airedYear, defaultValue),
                        DiscordHelpers.valueOrDefault(anime.season(), defaultValue)
                ), true)
                .addField(translate.t("CMD_ANIME_FIELD_STUDIOS"), studios, true);

        if (imageUrl != null && !imageUrl.isBlank()) {
            embed.image(imageUrl);
        }

        return embed.build();
    }
}
